python/pyarrow/src_arrow/CMakeLists.txt

```diff
@@ -0,0 +1,467 @@
+set(ARROW_PYTHON_SHARED_PRIVATE_LINK_LIBS)
+set(ARROW_PYTHON_STATIC_LINK_LIBS ${PYTHON_OTHER_LIBS})
+
+if(WIN32)
+  list(APPEND ARROW_PYTHON_SHARED_LINK_LIBS ${PYTHON_LIBRARIES} ${PYTHON_OTHER_LIBS})
+endif()
+if(PARQUET_REQUIRE_ENCRYPTION)
+  list(APPEND ARROW_PYTHON_SHARED_LINK_LIBS parquet_shared)
+endif()
+if(ARROW_USE_XSIMD)
+  list(APPEND ARROW_PYTHON_SHARED_PRIVATE_LINK_LIBS xsimd)
+  list(APPEND ARROW_PYTHON_STATIC_LINK_LIBS xsimd)
+endif()
+
+set(ARROW_PYTHON_INCLUDES ${NUMPY_INCLUDE_DIRS} ${PYTHON_INCLUDE_DIRS})
+
+# Inlude macros needed to find and use add_arrow_lib function
+include(BuildUtils)
+include(CMakePackageConfigHelpers)
+
+# Set the output directory for cmake module
+# (CMAKE_INSTALL_PREFIX = python/build/dist! should be set in setup.py!)
+set(ARROW_CMAKE_INSTALL_DIR "${CMAKE_INSTALL_LIBDIR}/cmake/${PROJECT_NAME}")
+
+add_arrow_lib(arrow_python
+              CMAKE_PACKAGE_NAME
+              ArrowPython
+              PKG_CONFIG_NAME
+              arrow-python
+              SOURCES
+              ${ARROW_PYTHON_SRCS}
+              PRECOMPILED_HEADERS
+              "$<$<COMPILE_LANGUAGE:CXX>:pch.h>"
+              OUTPUTS
+              ARROW_PYTHON_LIBRARIES
+              DEPENDENCIES
+              ${ARROW_PYTHON_DEPENDENCIES}
+              SHARED_LINK_FLAGS
+              ${ARROW_VERSION_SCRIPT_FLAGS} # Defined in line 95
+              SHARED_LINK_LIBS
+              ${ARROW_PYTHON_SHARED_LINK_LIBS}
+              SHARED_PRIVATE_LINK_LIBS
+              ${ARROW_PYTHON_SHARED_PRIVATE_LINK_LIBS}
+              STATIC_LINK_LIBS
+              ${ARROW_PYTHON_STATIC_LINK_LIBS}
+              EXTRA_INCLUDES
+              "${ARROW_PYTHON_INCLUDES}")
+
+add_dependencies(arrow_python ${ARROW_PYTHON_LIBRARIES})
+
+foreach(LIB_TARGET ${ARROW_PYTHON_LIBRARIES})
+  target_compile_definitions(${LIB_TARGET} PRIVATE ARROW_PYTHON_EXPORTING)
+endforeach()
+
+if(ARROW_BUILD_STATIC AND MSVC)
+  target_compile_definitions(arrow_python_static PUBLIC ARROW_STATIC)
+endif()
+
+if(ARROW_FLIGHT AND ARROW_BUILD_SHARED)
+
+  find_package(ArrowFlight REQUIRED)
+  find_package(gRPC CONFIG REQUIRED)
+  include_directories("${ARROW_SOURCE_DIR}/src" "${ARROW_SOURCE_DIR}/build/src")
+
+  # Using arrow_flight_shared, see example
+  # https://github.com/apache/arrow/blob/7a0f00c16e084d194ae53d209b33b809cfc8f2d5/cpp/examples/arrow/CMakeLists.txt
+  set(ARROW_GRPC_USE_SHARED ON)
+  set(GRPC_REFLECTION_LINK_LIBS -Wl,--no-as-needed gRPC::grpc++_reflection
+                                  -Wl,--as-needed)
+
+  set(FLIGHT_PROTO_PATH "${ARROW_SOURCE}/format")
+  set(FLIGHT_PROTO "${ARROW_SOURCE}/format/Flight.proto")
+
+  set(FLIGHT_GENERATED_PROTO_FILES
+      "${ARROW_SOURCE_DIR}/build/src/arrow/flight/Flight.pb.cc"
+      "${ARROW_SOURCE_DIR}/build/src/arrow/flight/Flight.pb.h"
+      "${ARROW_SOURCE_DIR}/build/src/arrow/flight/Flight.grpc.pb.cc"
+      "${ARROW_SOURCE_DIR}/build/src/arrow/flight/Flight.grpc.pb.h")
```

Comment:
Why do you want to use `build/...` in source directory instead of `${CMAKE_CURRENT_BINARY_DIR}` or something?

Reply:
AFAIU when building C PyArrow `${CMAKE_CURRENT_BINARY_DIR}` points to the build dir of the C PyArrow build which is in arrow/python. But here I need files from Arrow C++ build. I had no other idea than to use relative path like so.

I will try to compile Python Flight bindings without including some internal headers, as @pitrou suggested, and then maybe this will be redundant.

Reply:
The use of `ARROW_BUILD_DIR` CMake variable (that you suggested) made this nicer, I think.